malicious-site-protection/malicious-site-protection-impl/src/main/kotlin/com/duckduckgo/malicioussiteprotection/impl/data/MaliciousSiteRepository.kt

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2024 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.malicioussiteprotection.impl.data
+
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.malicioussiteprotection.impl.data.db.MaliciousSiteDao
+import com.duckduckgo.malicioussiteprotection.impl.data.network.FilterSetResponse
+import com.duckduckgo.malicioussiteprotection.impl.data.network.HashPrefixResponse
+import com.duckduckgo.malicioussiteprotection.impl.data.network.MaliciousSiteService
+import com.duckduckgo.malicioussiteprotection.impl.models.Feed.MALWARE
+import com.duckduckgo.malicioussiteprotection.impl.models.Feed.PHISHING
+import com.duckduckgo.malicioussiteprotection.impl.models.Filter
+import com.duckduckgo.malicioussiteprotection.impl.models.FilterSetWithRevision.MalwareFilterSetWithRevision
+import com.duckduckgo.malicioussiteprotection.impl.models.FilterSetWithRevision.PhishingFilterSetWithRevision
+import com.duckduckgo.malicioussiteprotection.impl.models.HashPrefixesWithRevision.MalwareHashPrefixesWithRevision
+import com.duckduckgo.malicioussiteprotection.impl.models.HashPrefixesWithRevision.PhishingHashPrefixesWithRevision
+import com.duckduckgo.malicioussiteprotection.impl.models.Match
+import com.duckduckgo.malicioussiteprotection.impl.models.Type
+import com.squareup.anvil.annotations.ContributesBinding
+import dagger.SingleInstanceIn
+import javax.inject.Inject
+import timber.log.Timber
+
+interface MaliciousSiteRepository {
+    suspend fun containsHashPrefix(hashPrefix: String): Boolean
+    suspend fun getFilters(hash: String): List<Filter>?
+    suspend fun matches(hashPrefix: String): List<Match>
+    suspend fun loadFilters()
+    suspend fun loadHashPrefixes()
+}
+
+@ContributesBinding(AppScope::class)
+@SingleInstanceIn(AppScope::class)
+class RealMaliciousSiteRepository @Inject constructor(
+    private val maliciousSiteDao: MaliciousSiteDao,
+    private val maliciousSiteService: MaliciousSiteService,
+) : MaliciousSiteRepository {
+
+    override suspend fun loadFilters() {
```

Comment:
We don't have any checks to avoid concurrency here, and ensure we are not updating db from multiple threads.

I would add a simple check to ensure if one job is running, we don't execute. (I believe we can use withContext for that as well)